src/common/subscribe.ts

```diff
@@ -0,0 +1,269 @@
+            }
+        }
+    }
+
+    return Promise.all(results).then((availables) => apiKeys.filter((_, index) => availables[index]))
+}
+
+export async function validateApiKey(url: string, apiKey: string, signal: AbortSignal): Promise<boolean> {
+    if (!apiKey || !url) {
+        return false
+    }
+
+    const headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${apiKey}` }
+    const body = JSON.stringify({
+        model: 'gpt-3.5-turbo',
+        max_tokens: 5,
+        stream: true,
+        messages: [
+            {
+                role: 'system',
+                content: 'You are a Python developer',
+            },
+            {
+                role: 'user',
+                content: 'Hello, world!',
+            },
+        ],
+    })
+
+    return await getUniversalFetch()(url, { method: 'POST', headers: headers, body: body, signal: signal })
+        .then((response) => (response?.status || 400) as number)
+        .then((status) => status === 200)
+        .catch(() => false)
+}
+
+async function generateAuthCodeModel(apiURL: string): Promise<AuthCodeModel | undefined> {
+    const payloads: Map<string, Record<string, string>> = new Map([
+        [`${apiURL}/api/user`, { authcode: '' } as Record<string, string>],
+        [`${apiURL}/api/models`, { key: '' } as Record<string, string>],
+    ])
+
+    const responses = await Promise.all(
+        Array.from(payloads).map(([url, payload]) => {
+            return new Promise((resolve) => {
+                ;(async () => {
+                    await getUniversalFetch()(url, {
+                        method: 'POST',
+                        headers: { 'Content-Type': 'application/json' },
+                        body: JSON.stringify(payload),
+                    })
+                        .then((response) => (response?.status === 200 ? response?.json() : undefined))
+                        .then(resolve)
+                        .catch(() => resolve(undefined))
+                })()
+            })
+        })
+    )
+
+    try {
+        const user = (responses[0] || {}) as Record<string, any>
+        const models = (responses[1] || []) as OpenAIModelDetail[]
+        const authCode: string = user?.authCode || ''
+
+        if (!authCode || typeof user.authCode !== 'string') {
+            return undefined
+        }
+
+        let model: OpenAIModelDetail = defaultOpenAIModelDetail
+        if (models) {
+            const modelIndexes = new Map<string, OpenAIModelDetail>()
+            for (let index = 0; index < models.length; index++) {
+                const item = models[index]
+                if (!item?.id) {
+                    continue
+                }
+                modelIndexes.set(item.id, item)
+            }
+
+            model = modelIndexes.get('gpt-4') || modelIndexes.get(defaultAPIModel) || models[0]
+        }
+
+        return { authCode: authCode, model: model }
+    } catch {
+        return undefined
```

Comment:
What error will be caught by this? Ignoring certain error messages may result in difficulties during debugging.

Reply:
Perhaps we can eliminate the try-catch block.

Reply:
There should be no errors theoretically, and I will remove it later.